File: Data_Structures/Stacks/stack_operations.c
/*
Write a program to implement a stack and perform the following operations:
- Push an element onto the stack
- Pop an element from the stack
- Find the minimum element in the stack in 0(1) time complexity
*/

#include <stdio.h>
#define MAX_SIZE 100

typedef struct {
    int stack[MAX_SIZE];
    int min_stack[MAX_SIZE];
    int top;
} Stack;

void initialize(Stack* stack) {
    stack->top = -1;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

int isFull(Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

void push(Stack* stack, int element) {
    if (isFull(stack)) {
        printf("Stack is full. Cannot push element.\n");
        return;
    }

    stack->stack[++stack->top] = element;

    if (isEmpty(stack) || element < stack->min_stack[stack->top - 1])
        stack->min_stack[stack->top] = element;
    else
        stack->min_stack[stack->top] = stack->min_stack[stack->top - 1];
}

int pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty. Cannot pop element.\n");
        return -1;
    }

    int element = stack->stack[stack->top--];
    stack->min_stack[stack->top + 1] = -1;

    return element;
}

int getMinimum(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty.\n");
        return -1;
    }

    return stack->min_stack[stack->top];
}

int main() {
    Stack stack;
    initialize(&stack);

    push(&stack, 5);
    push(&stack, 3);
    push(&stack, 8);
    push(&stack, 2);
    push(&stack, 6);

    printf("Minimum element: %d\n", getMinimum(&stack));

    pop(&stack);
    pop(&stack);

    printf("Minimum element: %d\n", getMinimum(&stack));

    printf("\n");
    return 0;
}

File: Data_Structures/Stacks/stack_operations.py
# Write a program to implement a stack and perform the following operations:
# - Push an element onto the stack
# - Pop an element from the stack
# - Find the minimum element in the stack in 0(1) time complexity

class Stack:
    def __init__(self):
        self.stack = []
        self.min_stack = []
    
    def is_empty(self):
        return len(self.stack) == 0
    
    def is_full(self):
        return False
    
    def push(self, element):
        self.stack.append(element)

        if self.is_empty() or element < self.get_minimum():
            self.min_stack.append(element)
        else:
            self.min_stack.append(self.min_stack[-1])
    
    def pop(self):
        if self.is_empty():
            print("Stack is empty. Cannot pop element.")
            return None
        
        self.min_stack.pop()
        return self.stack.pop()

    def get_minimum(self):
        if self.is_empty():
            print("Stack is empty.")
            return None
        
        return self.min_stack[len(self.min_stack) - 1]

stack = Stack()

stack.push(5)
stack.push(3)
stack.push(8)
stack.push(2)
stack.push(6)

print("Minimum element:", stack.get_minimum())

stack.pop()
stack.pop()

print("Minimum element:", stack.get_minimum())
